libraries/ESP32/examples/TWAI/TWAIreceive/TWAIreceive.ino

```diff
@@ -0,0 +1,111 @@
+  -> RX buffered: 4  RX missed: 46 RX overrun 0
+
+  created 05-11-2022 by Stephan Martin (designer2k2)
+*/
+
+#include "driver/twai.h"
+
+//Pins used to connect to CAN bus transceiver:
+#define rxPin 21
+#define txPin 22
+
+void setup() {
+  // Start Serial:
+  Serial.begin(115200);
+
+  //Initialize configuration structures using macro initializers
+  twai_general_config_t g_config = TWAI_GENERAL_CONFIG_DEFAULT((gpio_num_t)txPin, (gpio_num_t)rxPin, TWAI_MODE_LISTEN_ONLY);
+  twai_timing_config_t t_config = TWAI_TIMING_CONFIG_500KBITS();  //Look in the api-reference for other speed sets.
+  twai_filter_config_t f_config = TWAI_FILTER_CONFIG_ACCEPT_ALL();
+
+  //Install TWAI driver
+  if (twai_driver_install(&g_config, &t_config, &f_config) == ESP_OK) {
+    Serial.println("Driver installed");
+  } else {
+    Serial.println("Failed to install driver");
+    return;
+  }
+
+  //Start TWAI driver
+  if (twai_start() == ESP_OK) {
+    Serial.println("Driver started");
+  } else {
+    Serial.println("Failed to start driver");
+    return;
+  }
+
+  //Reconfigure alerts to detect Bus-Off error and RX queue full states
+  uint32_t alerts_to_enable = TWAI_ALERT_ERR_PASS | TWAI_ALERT_BUS_ERROR | TWAI_ALERT_RX_QUEUE_FULL;
+  if (twai_reconfigure_alerts(alerts_to_enable, NULL) == ESP_OK) {
+    Serial.println("CAN Alerts reconfigured");
+  } else {
+    Serial.println("Failed to reconfigure alerts");
+  }
+
+}
+
+void loop() {
+  //Check if alert happened
+  uint32_t alerts_triggered;
+  twai_read_alerts(&alerts_triggered, pdMS_TO_TICKS(1));
+  twai_status_info_t twaistatus;
+  twai_get_status_info(&twaistatus);
+  if (alerts_triggered & TWAI_ALERT_ERR_PASS) {
+    Serial.println("Alert: TWAI controller has become error passive.");
+  }
+  if (alerts_triggered & TWAI_ALERT_BUS_ERROR) {
+    Serial.println("Alert: A (Bit, Stuff, CRC, Form, ACK) error has occurred on the bus.");
+    Serial.printf("Bus error count: %d\n", twaistatus.bus_error_count);
+  }
+  if (alerts_triggered & TWAI_ALERT_RX_QUEUE_FULL) {
+    Serial.println("Alert: The RX queue is full causing a received frame to be lost.");
+    Serial.printf("RX buffered: %d\t", twaistatus.msgs_to_rx);
+    Serial.printf("RX missed: %d\t", twaistatus.rx_missed_count);
+    Serial.printf("RX overrun %d\n", twaistatus.rx_overrun_count);
+  }
+
+  //Wait for message to be received:
+  twai_message_t message;
+  if (twai_receive(&message, pdMS_TO_TICKS(1000)) == ESP_OK) {
+    Serial.println("Message received");
+  } else {
+    Serial.println("No message received");
+    return;
+  }
+
+  //Process received message
+  if (message.extd) {
+    Serial.println("Message is in Extended Format");
+  } else {
+    Serial.println("Message is in Standard Format");
+  }
+  Serial.printf("ID is %x\n", message.identifier);
+  if (!(message.rtr)) {
+    for (int i = 0; i < message.data_length_code; i++) {
+      Serial.printf("Byte %d = %02x ", i, message.data[i]);
+    }
+    Serial.println("");
+  }
+
+}
```

Comment:
Might be a good idea to use the `TWAI_ALERT_RX_DATA`. It was specifically added ([commit](https://github.com/espressif/esp-idf/commit/72ba1866fbd2f3eb91754c01d1f9aa23f69495cb), [issue](https://github.com/espressif/esp-idf/issues/7374)) to allow for polling usage of the TWAI driver (opposed to event driven).

This way, the `twai_read_alerts()` acts like an event queue, and we can run the loop at a fixed polling rate demonstrated below. IMO this is cleaner as we only block once in each loop iteration.

```suggestion
void loop() {
  uint32_t alerts_triggered;
  twai_read_alerts(&alerts_triggered, pdMS_TO_TICKS(POLLING_RATE_MS));
  ...
  // Handle alerts
  if (alerts_triggered & TWAI_ALERT_RX_DATA) {
    // One or more messages received. Clear all
    twai_message_t message;
    while (twai_receive(&message, 0) == ESP_OK) {
        ...
    }
  }
  if (alerts_triggered & TWAI_ALERT_TX_IDLE) {
    /* If we want to demonstrate how to transmit as well, TWAI_ALERT_TX_IDLE can be used to indicate when
    all previous twai_transmit() messages have completed. Thus, we can start loading more messages to transmit here.*/
  }
  if (alerts_triggered & TWAI_ALERT_ERR_PASS) {
    ...
  }
  if (alerts_triggered & TWAI_ALERT_BUS_ERROR) {
    ...
  }
  if (alerts_triggered & TWAI_ALERT_RX_QUEUE_FULL) {
    ...
  }
  
}
```

Reply:
i added the TWAI_ALERT_RX_DATA but left out the TX part, that is then for a send example.